aten/src/ATen/native/Distributions.h

```diff
@@ -0,0 +1,215 @@
+// is adapted from Numpy's distributions.c implementation.
+// It is MIT licensed, so here is the copyright:
+
+/* Copyright 2005 Robert Kern (robert.kern@gmail.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+template<typename scalar_t>
+deviceforcuda scalar_t sample_gamma(scalar_t alpha, BaseSampler<scalar_t>& standard_uniform, BaseSampler<scalar_t>& standard_normal) {
+  scalar_t scale = 1.0;
+
+  // Boost alpha for higher acceptance probability.
+  if (alpha < 1.0) {
+    scale *= std::pow(1 - standard_uniform.sample(), static_cast<scalar_t>(1.0) / alpha);
+    alpha += 1.0;
+  }
+
+  // This implements the acceptance-rejection method of Marsaglia and Tsang (2000)
+  // doi:10.1145/358407.358414
+  const scalar_t d = alpha - 1.0 / 3.0;
+  const scalar_t c = 1.0 / std::sqrt(9.0 * d);
+  for (;;) {
+    scalar_t x, y;
+    do {
+      x = standard_normal.sample();
+      y = 1.0 + c * x;
+    } while (y <= 0);
+    const scalar_t v = y * y * y;
+    const scalar_t u = 1 - standard_uniform.sample();
+    const scalar_t xx = x * x;
+    if (u < 1.0 - 0.0331 * xx * xx)
+      return scale * d * v;
+    if (std::log(u) < 0.5 * xx + d * (1.0 - v + std::log(v)))
+      return scale * d * v;
+  }
+}
+
+template <typename scalar_t>
+deviceforcuda static inline scalar_t polevl(const scalar_t x,  const scalar_t A[], size_t len) {
+  scalar_t result = 0;
+  for (size_t i = 0; i <= len; i++) {
+    result = result * x + A[i];
+  }
+  return result;
+}
+
+
+/*
+ * The following function comes with the following copyright notice.
+ * It has been released under the BSD license.
+ *
+ * Cephes Math Library Release 2.8:  June, 2000
+ * Copyright 1984, 1987, 1992, 2000 by Stephen L. Moshier
+ */
+template <typename scalar_t>
+deviceforcuda static inline scalar_t digamma_one(scalar_t x) {
+  using acc_scalar_t = typename std::conditional<std::is_same<scalar_t, at::Half>::value, float, scalar_t>::type;
```

Comment:
How can I use `using accscalar_t = cuda::acc_type<scalar_t>;` when want the same code on the CPU?

Reply:
We should probably get `acc_type` that doesn't really depend on `cuda`

Reply:
I think it requires some decision to be made, because currently in TH accumulate type for float is double, whereas in THC accumulate type for float is float, I think there also some differences in integer accumulate types. If we want to mimic this behavior, we cannot have an accumulate type that's independent of cuda.